Tray notifications (system native notifications)

Delwing opened the pull request:
<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions
Some messages are more important than others, so important that they should show up even when Mudlet is minimized! We can utilize system native notifications for that.

![image](https://user-images.githubusercontent.com/3740628/101994450-da605380-3cc2-11eb-9333-bf2874841dd6.png)
![image](https://user-images.githubusercontent.com/3740628/101995808-16001b00-3ccd-11eb-8d1d-1e0985d522a9.png)



#### Motivation for adding to Mudlet
Pretty straightforward API that can display toast/baloon notifications using native system notification system,

#### Other info (issues closed, discussion etc)
Availability of notifications might vary from system to system.
Especially timeout parameter might be ignored by system.
Notification themselves can be not displayed if system is configured that way, but for those with desire to see notifications from Mudlet will work like charm.

Commit:
add tray icon for mudlet

Commit:
add lua api for showing native desktop notifications

Commit:
add trailing new line for mudlet.cpp

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17924,36 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
 }
+
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#showNativeNotification
+int TLuaInterpreter::showNativeNotification(lua_State* L)
```
I think the word 'native' will be lost on most people. Just `showNotification` would be OK.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -661,6 +661,8 @@ mudlet::mudlet()
     // the current GUI Language
     loadMaps();
+
+    setupTrayIcon();
```
I think a fair bit of people will find the intrusion into their notification area annoying, esp. since most scripts won't use a notification for a while. How about setting it up automatically as soon as a notification should be shown?

This does leave further questions as to when to hide it but I think we can leave it be once shown.

Commit:
fix included added in ifdef

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17924,36 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
+{
+    int n = lua_gettop(L);
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "showNativeNotification: bad argument #1 type (word as string expected, got %s!)", luaL_typename(L, 1));
```
```suggestion
        lua_pushfstring(L, "showNativeNotification: bad argument #1 type (title as string expected, got %s!)", luaL_typename(L, 1));
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17924,36 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
+        return lua_error(L);
+    }
+    if (n >= 2 && !lua_isstring(L, 2)) {
+        lua_pushfstring(L, "showNativeNotification: bad argument #2 type (word as string expected, got %s!)", luaL_typename(L, 2));
```
```suggestion
        lua_pushfstring(L, "showNativeNotification: bad argument #2 type (text as string expected, got %s!)", luaL_typename(L, 2));
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17924,36 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
+    }
+    int notificationExpirationTime = 1000;
+    if (n >= 3 && !lua_isnumber(L, 3)) {
+        lua_pushfstring(L, "showNativeNotification: bad argument #3 type (number expected, got %s!)", luaL_typename(L, 3));
```
```suggestion
        lua_pushfstring(L, "showNativeNotification: bad argument #3 type (expiration time as number expected, got %s!)", luaL_typename(L, 3));
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17924,36 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
+        lua_pushfstring(L, "showNativeNotification: bad argument #3 type (number expected, got %s!)", luaL_typename(L, 3));
+        return lua_error(L);
+    } else if (lua_isnumber(L, 3)) {
+        notificationExpirationTime = lua_tonumber(L, 3);
```
It would be good for this to be in seconds, more natural to most people.

vadi2 reviewed (commented):
Lookin good!

Commit:
change showNativeNotification to showNotification
improve error messages
make notification expiration to take value in seconds
show tray icon only after first notification sent

Delwing commented:
Corrections pushed.

vadi2 reviewed (approved):
Looks good. One final idea to address the mandatory-for-notifications system tray: what about also adding a 'Hide tray' option into it, so if someone wants to, they can hide it by hand?

Delwing commented:
I like to rely on system to handle things like that. That way there is one, uniform way to handle all tray icons etc. Although since it serves no bigger purpose I might add add either auto hide after certain time or just in context menu, but it will appear every time notification is sent.

vadi2 commented:
Not to the notification - the system tray, to hide it when unwanted :)

Delwing commented:
I was not referring to notification but tray icon :) For notification to be visible, tray icon must be visible as well, therefore after your suggestion I am calling show before showing notification.

Delwing commented:
https://wiki.mudlet.org/w/Manual:Miscellaneous_Functions#showNotification docs added.
I guess rebuild is required.

So do you think some additional context menu for hiding would be required, since it will go back to tray again with each new `showMessage`

vadi2 commented:
Yeah it would, but if you aren't getting any new notifications for a while, might be nice to be able to hide it. It's not a dealbreaker, let's see how it works in experience and whenever we need to tweak it further.

vadi2 commented on src/mudlet.h:
```diff
@@ -384,6 +386,7 @@ class mudlet : public QMainWindow, public Ui::main_window
     QTextOption::Flags mEditorTextOptions;
 
     QPointer<QTableWidget> moduleTable;
+    QSystemTrayIcon mTrayIcon = QSystemTrayIcon(QIcon(":/icons/mudlet.png"));
```
Maybe you can make the dev and ptb versions use different icons as they do already?

People on Linux will want to be able to customise the icon to be monochrome or something else, that would be nice sugar.

Delwing replied:
At the moment I think I found quite elegant solution - tray icon will be taken directly from mudlet object.

vadi2 replied:
I agree, very nice.

Commit:
make tray icon to be the same as window icon

Commit:
add hide tray icon action for tray icon menu
